# Project: Wrangling and Analyze Data

## Data Gathering
In the cell below, gather **all** three pieces of data for this project and load them in the notebook. **Note:** the methods required to gather each data are different.
1. Directly download the WeRateDogs Twitter archive data (twitter_archive_enhanced.csv)

In [3]:
# import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import json
import requests

%matplotlib inline

In [4]:
# load the twitter-archive-enhanced.csv into a dataframe
twit_arch = pd.read_csv('twitter-archive-enhanced.csv')

2. Use the Requests library to download the tweet image prediction (image_predictions.tsv)

In [5]:
# downlod the image prediction file programmatically
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)

with open(url.split('/')[-1], mode = 'wb') as file:
    file.write(response.content)

In [6]:
# load the image-prediction.tsv into a dataframe
image_pred = pd.read_table('image-predictions.tsv')

3. Use the Tweepy library to query additional data via the Twitter API (tweet_json.txt)

In [7]:
# I have no access to twitter API, so download the tweet-json provided by Udacity
data = []
with open('tweet-json.txt', 'r')  as file:
    for line in file:
        data.append(json.loads(line))

In [8]:
# extract the contents needed(tweet_id, retweet_count, favorite_count)
deets = []
for item in data:
    tweet_id = item['id']
    retweet_count = item['retweet_count']
    favorite_count = item['favorite_count']
    
    deets.append({'tweet_id': tweet_id,
                 'retweet_count': retweet_count,
                 'favorite_count': favorite_count})

In [9]:
# read the extracted content into a dataframe
twit_like = pd.DataFrame(deets, columns = ['tweet_id', 'retweet_count', 'favorite_count'])

## Assessing Data
In this section, detect and document at least **eight (8) quality issues and two (2) tidiness issue**. You must use **both** visual assessment
programmatic assessement to assess the data.

**Note:** pay attention to the following key points when you access the data.

* You only want original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.
* Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.
* The fact that the rating numerators are greater than the denominators does not need to be cleaned. This [unique rating system](http://knowyourmeme.com/memes/theyre-good-dogs-brent) is a big part of the popularity of WeRateDogs.
* You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.



In [10]:
# First assess the tables (twitter-archive-enhanced) visually with google sheets, or any other program, for review, I'll use pandas .sample() function
twit_arch.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
215   851224888060895234                    NaN                  NaN   
226   848690551926992896                    NaN                  NaN   
500   813112105746448384                    NaN                  NaN   
1572  687664829264453632                    NaN                  NaN   
1649  683742671509258241                    NaN                  NaN   

                      timestamp  \
215   2017-04-10 00:06:42 +0000   
226   2017-04-03 00:16:10 +0000   
500   2016-12-25 20:00:07 +0000   
1572  2016-01-14 15:57:26 +0000   
1649  2016-01-03 20:12:10 +0000   

                                                 source  \
215   <a href="http://twitter.com/download/iphone" r...   
226   <a href="http://twitter.com/download/iphone" r...   
500   <a href="http://twitter.com/download/iphone" r...   
1572  <a href="http://twitter.com/download/iphone" r...   
1649  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
215   Meet Snoop. His number one passion is sticking...                  NaN   
226   Please stop sending in animals other than dogs...                  NaN   
500   Meet Toby. He's pupset because his hat isn't b...                  NaN   
1572  Meet Opal. He's a Belgian Dijon Poofster. Upse...                  NaN   
1649  Meet Sebastian. He's a womanizer. Romantic af....                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
215                        NaN                        NaN   
226                        NaN                        NaN   
500                        NaN                        NaN   
1572                       NaN                        NaN   
1649                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
215   https://twitter.com/dog_rates/status/851224888...                13   
226   https://twitter.com/dog_rates/status/848690551...                12   
500   https://twitter.com/dog_rates/status/813112105...                12   
1572  https://twitter.com/dog_rates/status/687664829...                11   
1649  https://twitter.com/dog_rates/status/683742671...                11   

      rating_denominator       name doggo floofer pupper puppo  
215                   10      Snoop  None    None   None  None  
226                   10       None  None    None   None  None  
500                   10       Toby  None    None   None  None  
1572                  10       Opal  None    None   None  None  
1649                  10  Sebastian  None    None   None  None

In [11]:
# Assess the tables(image-prediction) visually
image_pred.sample(5)

tweet_id                                          jpg_url  \
653   682047327939461121  https://pbs.twimg.com/media/CXcebTeWsAUQJ-J.jpg   
1495  783347506784731136  https://pbs.twimg.com/media/CVuQ2LeUsAAIe3s.jpg   
992   708109389455101952  https://pbs.twimg.com/media/CdO1u9vWAAApj2V.jpg   
351   672523490734551040  https://pbs.twimg.com/media/CVVIjGbWwAAxkN0.jpg   
900   699801817392291840  https://pbs.twimg.com/media/CbYyCMcWIAAHHjF.jpg   

      img_num                         p1   p1_conf  p1_dog  \
653         1                      teddy  0.364095   False   
1495        1                   Cardigan  0.611525    True   
992         1  Staffordshire_bullterrier  0.516106    True   
351         1           golden_retriever  0.565981    True   
900         2           golden_retriever  0.808978    True   

                                  p2   p2_conf  p2_dog                  p3  \
653                          doormat  0.119243   False             toyshop   
1495                        Pembroke  0.368566    True           Chihuahua   
992   American_Staffordshire_terrier  0.236075    True              kelpie   
351                             chow  0.081212    True       Irish_terrier   
900                     Irish_setter  0.042428    True  Labrador_retriever   

       p3_conf  p3_dog  
653   0.035127   False  
1495  0.003330    True  
992   0.069750    True  
351   0.061596    True  
900   0.023536    True

In [12]:
# Assess the tables(retweet and tweet likes) visually
twit_like.sample(5)

tweet_id  retweet_count  favorite_count
133   866686824827068416           3727           20070
826   768970937022709760           7574           16017
2215  668528771708952576            242             496
1331  705442520700944385           1859            4877
1734  679722016581222400            539            1817

Assess each table programatically with pandas methods or functions(.head(), .describe(), .duplicated(), .info(), e.t.c)

In [13]:
# Assess each table programatically (twitter-archive-enhanced)
twit_arch.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [14]:
# show statistical summary of table
twit_arch.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [18]:
# check number of duplicated rows
sum(twit_arch.duplicated())

0

In [21]:
# show number of missing values in each column
twit_arch.isna().sum()

tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

In [22]:
# Assess each table programatically (image-prediction)
image_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [23]:
# show statistical summary of table
image_pred.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [24]:
# check number of duplicated rows
sum(image_pred.duplicated())

0

In [25]:
# Assess each table programatically (tweet-json)
twit_like.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2354 non-null   int64
 1   retweet_count   2354 non-null   int64
 2   favorite_count  2354 non-null   int64
dtypes: int64(3)
memory usage: 55.3 KB


In [26]:
# show statistical summary of table
twit_like.describe()

tweet_id  retweet_count  favorite_count
count  2.354000e+03    2354.000000     2354.000000
mean   7.426978e+17    3164.797366     8080.968564
std    6.852812e+16    5284.770364    11814.771334
min    6.660209e+17       0.000000        0.000000
25%    6.783975e+17     624.500000     1415.000000
50%    7.194596e+17    1473.500000     3603.500000
75%    7.993058e+17    3652.000000    10122.250000
max    8.924206e+17   79515.000000   132810.000000

In [27]:
# check number of duplicated rows
sum(twit_like.duplicated())

0

### Quality issues
1. twitter-archive-enhanced table: Tons of missing values in the table.

2. twitter-archive-enhanced table: Columns that are not relevant to analysis are present

3. twitter-archive-enhanced table: incorrect datatype for timestamp column

4. image-preiction table: missing records leading to incomplete dataset (2075 out of 2356)

5. twitter-archive-enhanced table: Null values represented as None in name column

6. image-preiction table: names of columns are not explanatory

7. twitter-archive-enhanced table: maximum and minimum value for rating_denomenator column are 170 and 0 instead of 10

8. twitter-archive-enhanced table: p1, p2 and p3 column values containing underscore(_)

### Tidiness issues
1. twitter-archive-enhanced table: Dog stages are divided into columns 

2. twitter-archive-enhanced table: two variables of date and time in the timestamp column

## Cleaning Data
In this section, clean **all** of the issues you documented while assessing. 

**Note:** Make a copy of the original data before cleaning. Cleaning includes merging individual pieces of data according to the rules of [tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html). The result should be a high-quality and tidy master pandas DataFrame (or DataFrames, if appropriate).

In [58]:
# Make copies of original pieces of data
twit_arch_cp = twit_arch.copy()
image_pred_cp = image_pred.copy()
twit_like_cp = twit_like.copy()

### Issue #1:

#### Define:

#### Code

#### Test

### Issue #2:

#### Define

#### Code

#### Test

## Storing Data
Save gathered, assessed, and cleaned master dataset to a CSV file named "twitter_archive_master.csv".

## Analyzing and Visualizing Data
In this section, analyze and visualize your wrangled data. You must produce at least **three (3) insights and one (1) visualization.**

### Insights:
1.

2.

3.

### Visualization